shared/js/background/classes/sw-tab.es6.js

```diff
@@ -0,0 +1,62 @@
+const Tab = require('./tab.es6')
+
+/**
+ * A stub tab implementation for service workers.
+ *
+ * For most actions, this just mimics a tab with the service worker URL as the tab URL. However, for cases
+ * such as tracker counts, it distributes stats to all known tabs with the same origin.
+ */
+class ServiceWorkerTab extends Tab {
+    /**
+     * @param {string} swUrl
+     * @param {Record<number, Tab>} tabContainer
+     */
+    constructor (swUrl, tabContainer) {
+        super({
+            tabId: -1,
+            url: swUrl,
+            status: null
+        })
+        this.origin = new URL(swUrl).origin
+        this.tabContainer = tabContainer
+    }
+
+    /**
+     * Find the list of tabs which share the same origin as this service worker.
+     * @returns {Tab[]}
+     */
+    _findMatchingTabs () {
```

Comment:
I'm figuring this will be called quite often, like for every request with a tabId of -1. So perhaps we should have a more efficient way of finding the relevant tabs?

How about we give ServiceWorker tabs a WeakMap of associated Tab Objects (same origin/entity). Each time a normal Tab Object is created, we update the WeakMaps for all the ServiceWorker Tabs. Then this function could just return the values of the WeakMap. WDYT?

Reply:
So you're suggesting to keep a live weak mapping of origin -> tab (updating on every webnavigation event), in order to make this lookup faster? I'd imagine there's be a load of edge cases to handle there (e.g. with extension SW restarts).

I'd argue that this is fast enough - we're just doing one loop through the tabs (probably N<100 for most cases), and only when there is a serviceWorker initiated request that is also a tracker (not common).

Reply:
Well, yea. I'm not set on the WeakMap approach. Another approach might be to maintain an origin to Tab mapping for efficient lookup. But parsing the URL and comparing origin for every tab, every time there's one of these requests does make me nervous. Some users have a lot of tabs. Maybe there's more of these requests than we expect. I guess it's not a problem until it is...

I suppose we can optimise it in the future too if necessary. How about adding a comment here at least? I figure it could explain that for large numbers of tabs + Serviceworker initiated tracking requests this could get slow. In the future we might want to consider maintaining a more efficient way to find Tab Objects that correspond to a given origin.

Reply:
I think this is premature optimization - the current implementation is `number of tracker requests via SW x number of tabs` `URL` parses for a page load, which is max in the order of 1000s. Note, we already parse the site domain at least once for every request, which is a much hotter path than this.

Additionally, sketching out the alternative, it sounds potentially tricky to implement, particularly with our tab (de)serialization going on (in MV3):
-  Create a mapping of `Map<string, WeakSet<Tab>>`
- Whenever we create a tab object, insert it into the map using the URL origin as the key.
- Also handle tab restores (`tabManager.restore` and `tabManager.restoreOrCreate`).
- Check that a tab's `url` is immutable. If not, handle modification (remove from one weak set and add to another).
- Cleanup origins with no more tabs (otherwise this will leak memory). Avoid doing this too often, as we have to iterate the map values, which will be almost as large as the total number of tabs.

Reply:
Alright, fair enough. I would have liked a short comment to note the cost and potential future improvement, but I don't insist. Up to you :+1: